FEXCore/Source/Interface/Core/OpcodeDispatcher/X87.cpp

```diff
@@ -970,43 +553,52 @@ void OpDispatchBuilder::X87FNSAVE(OpcodeArgs) {
   // ST7 broken in to two parts
   // Lower 64bits [63:0]
   // upper 16 bits [79:64]
-  _StoreMem(FPRClass, 8, data, Mem, _Constant((Size * 7) + (7 * 10)), 1, MEM_OFFSET_SXTX, 1);
+  _StoreMem(FPRClass, 8, ST0Location, data, 1);
+  ST0Location = _Add(OpSize::i64Bit, ST0Location, _Constant(8));
   auto topBytes = _VDupElement(16, 2, data, 4);
-  _StoreMem(FPRClass, 2, topBytes, Mem, _Constant((Size * 7) + (7 * 10) + 8), 1, MEM_OFFSET_SXTX, 1);
+  _StoreMem(FPRClass, 2, ST0Location, topBytes, 1);
 
   // reset to default
   FNINIT(Op);
 }
 
 void OpDispatchBuilder::X87FRSTOR(OpcodeArgs) {
+  CurrentHeader->HasX87 = true;
+  _StackForceSlow();
   const auto Size = GetSrcSize(Op);
   Ref Mem = MakeSegmentAddress(Op, Op->Src[0]);
 
   auto NewFCW = _LoadMem(GPRClass, 2, Mem, 2);
   _StoreContext(2, GPRClass, NewFCW, offsetof(FEXCore::Core::CPUState, FCW));
 
-  auto NewFSW = _LoadMem(GPRClass, Size, Mem, _Constant(Size * 1), Size, MEM_OFFSET_SXTX, 1);
-  auto Top = ReconstructX87StateFromFSW(NewFSW);
+  Ref MemLocation = _Add(OpSize::i64Bit, Mem, _Constant(Size * 1));
+  auto NewFSW = _LoadMem(GPRClass, Size, MemLocation, Size);
+  Ref Top = ReconstructX87StateFromFSW_Helper(NewFSW);
   {
     // FTW
-    SetX87FTW(_LoadMem(GPRClass, Size, Mem, _Constant(Size * 2), Size, MEM_OFFSET_SXTX, 1));
+    Ref MemLocation = _Add(OpSize::i64Bit, Mem, _Constant(Size * 2));
+    SetX87FTW(_LoadMem(GPRClass, Size, MemLocation, Size));
```

Comment:
Why are we changing this? I specifically folded in the Adds into these instructions a few months ago, this is a functional revert of that. Why?

Reply:
I went through your commit at the time and found many more places where this happened, so I fixed those as well. :-(